source/agora/network/Crawler.d

```diff
@@ -0,0 +1,570 @@
+        foreach (i, ref field; crawl_result.tupleof)
+            static if (i == 6)
+                break;
+            else static if (is(typeof(field) == string))
+                field = extracted_values[i];
+            else static if (is(typeof(field) == TimePoint))
+                field = this.clock.networkTime();
+            else static assert(0);
+
+        crawl_result.os = node_info.os;
+        crawl_result.client_name = node_info.client_name;
+        crawl_result.client_ver = node_info.client_ver;
+        crawl_result.height = node_info.height;
+        crawl_result.is_ipv4 = is_ipv4;
+
+        return crawl_result;
+    }
+
+    unittest
+    {
+        import std.algorithm : map;
+        import std.array : array;
+        import std.conv : to;
+
+        auto mock_clock = new MockClock(5);
+        auto crawler = new Crawler(null, mock_clock, Config(), null);
+        NodeInfo node_info;
+        node_info.os = "6";
+        node_info.client_name = "7";
+        node_info.client_ver = "8";
+        node_info.height = Height(9);
+
+        CrawlResult expected;
+        expected.continent = "0";
+        expected.country = "1";
+        expected.city = "2";
+        expected.latitude = "3";
+        expected.longitude = "4";
+        expected.crawl_time = 5;
+        expected.os = "6";
+        expected.client_name = "7";
+        expected.client_ver = "8";
+        expected.height = Height(9);
+        expected.is_ipv4 = true;
+
+        assert(crawler.populateCrawlResult(iota(5).map!(i => i.to!string()).array(), node_info, true) == expected);
+    }
+
+/***************************************************************************
+
+        Gets the information about the network and it's reachable nodes.
+
+        The information includes the network address and the geograpical location
+        of the nodes including the continent/country/city/latitude/longitude.
+
+        API:
+            GET /network_info
+
+        Returns:
+            information about the network and about it's nodes
+
+
+    ***************************************************************************/
+
+
+    public CrawlResultHolder getNetworkInfo () @safe @nogc nothrow pure
+    {
+        return crawl_result_holder;
+    }
+
+    /***************************************************************************
+
+        Creates and returns a newly created NodeLocator object.
+
+        Returns:
+            a newly created NodeLocator object
+
+    ***************************************************************************/
+
+    protected INodeLocator makeNodeLocator ()
+    {
+        return new NodeLocatorGeoIP(this.config.node.ipdb_path);
+    }
+
+    // Start the Crawler
+    public void start ()
+    {
+        import std.array : array;
+
+        if (!this.config.node.collect_network_statistics)
+            return;
+
+        // at least 1 seed node needs to be provided
+        this.network_manager.seed_addresses.length || assert(0);
```

Comment:
I suppose you mean the name-registry not the test registry. 

The crawler relies on the seed address to initiate the crawling. In theory we could have a 

- VALIDATOR
- that has no seeds configured
- we also want to use it as a network crawler

but it is kind of a corner case. I can modify the code though.

Reply:
I was thinking a bit more about this: I think we should keep the assert here, so our crawler can always recover from network outages. 

If we allow the following node to act as a crawler:

- Validator
- that has no seeds configured
- we also want to use it as a network crawler 

then what happens if the crawler node has a network outage, and then that outage causes every node in its BFS search tree to be (permanently) removed. Normally that search tree would never be empty, because we have seed nodes(although this node didn't define any). 

We shouldn't treat nodes that were returned from the name registry as seed nodes I think because those nodes can come and go. And if we really want to treat name regitsty nodes as seeds, then it should be done in the Network registry.